src/common/cluster/request-api-resources.injectable.ts

```diff
@@ -24,58 +25,49 @@ const requestApiResourcesInjectable = getInjectable({
-          (async () => {
-            const { versions } = await k8sRequest(cluster, "/api") as V1APIVersions;
+    const requestApiVersions = async (cluster: Cluster) => (await k8sRequest(cluster, "/api") as V1APIVersions).versions;
+    const requestApisVersions = async (cluster: Cluster) => (await k8sRequest(cluster, "/apis") as V1APIGroupList).groups;
+    const limitingFor = (limit: plimit.Limit) => <Args extends any[], Res>(fn: (...args: Args) => Res) => (...args: Args) => limit(() => fn(...args));
+    const requestKubeApiResourcesFor = (cluster: Cluster) => async ({ group, path }: KubeResourceListGroup): Promise<KubeApiResource[]> => {
+      const { resources } = await k8sRequest(cluster, path) as V1APIResourceList;
 
-            for (const version of versions) {
-              resourceListGroups.push({
-                group: version,
-                path: `/api/${version}`,
-              });
-            }
-          })(),
-          (async () => {
-            const { groups } = await k8sRequest(cluster, "/apis") as V1APIGroupList;
-
-            for (const { preferredVersion, name } of groups) {
-              const { groupVersion } = preferredVersion ?? {};
+      return resources.map(resource => ({
+        apiName: resource.name,
+        kind: resource.kind,
+        group,
+        namespaced: resource.namespaced,
+      }));
+    };
 
-              if (groupVersion) {
-                resourceListGroups.push({
-                  group: name,
-                  path: `/apis/${groupVersion}`,
-                });
-              }
-            }
-          })(),
-        ]);
+    return async (cluster) => {
+      const requestKubeApiResources = requestKubeApiResourcesFor(cluster);
+      const withApiLimit = limitingFor(plimit(5));
 
-        await Promise.all(
-          resourceListGroups.map(({ group, path }) => apiLimit(async () => {
-            const { resources } = await k8sRequest(cluster, path) as V1APIResourceList;
+      try {
+        const resourceListGroups: KubeResourceListGroup[] = [
+          ...(await requestApiVersions(cluster))
+            .map(version => ({
+              group: version,
+              path: `/api/${version}`,
+            })),
+          ...pipeline((await requestApisVersions(cluster)).values())
+            .filterMap(group => group.preferredVersion?.groupVersion && ({
+              group: group.name,
+              path: `/apis/${group.preferredVersion.groupVersion}`,
+            })),
+        ];
 
-            for (const resource of resources) {
-              kubeApiResources.push({
-                apiName: resource.name,
-                kind: resource.kind,
-                group,
-                namespaced: resource.namespaced,
-              });
-            }
-          })),
+        const resources = await Promise.all(
+          resourceListGroups
+            .map(withApiLimit(requestKubeApiResources)),
         );
+
+        return resources.flat();
```

Comment:
Could this be even flatter, consistent and more SRP with:

```
const resources = await pipeline(resourceListGroups)
  .map(toCallForKubeApiResources)
  .map(withApiLimit)
  // Note: this could be .callAll() in iter.
  .map(call => call())
  // Not sure if these exist in iter yet?
  .awaitAll()
  .flat()
  .asArray();
```

Reply:
Given that `awaitAll` doesn't exist. I think that just splitting the maps would be good enough to make this flat.

Reply:
I attempted that and the typings are not working out. Will push what I have so far since I think it is better. But I don't think that flatness is a quality indicator that must be pursued forever.

Reply:
Note: further pursuit will probably reveal some pretty cool stuff. Let's do that later, when time comes :)